### ISSUE TITLE
Feature request: Glossary

### ISSUE THREAD

**Issue by debruine** [quarto-dev/quarto-cli#1697]
I'd like to suggest a first-class glossary in quarto, where you can link terms in the text to definitions (perhaps displayed on hover) and collect defined terms into an optional glossary table or section at the end of a chapter or the end of a book.

See this [twitter thread](https://twitter.com/LisaDeBruine/status/1555543054974599169) for discussion about this with Thomas Mock and the [psyTeachR glossary package](https://psyteachr.github.io/glossary/) for a model of what I'm thinking about, implemented in R.

## Some ideas

It could require defining terms somehow in the text each time, like:

`{{< gls [term](This is my definition.) >}}`

Or there could be a glossary file similar to `_variable.yml` that you can use to link terms to:

`{{< gls [text-term](file-term) >}}`

This letter option might be better, as you often want to link the same term several times, or link different forms (e.g., the definition might be for "argument", but your sentence is "There are several arguments you need to set.").

**Comment by jthomasmock** [comment]
I agree that some inspiration for an external YAML file, a la `_pkgdown.yml` implementation might work here. You’re going to write lots of other documents/shortcodes/definitions but can control the organization of combined glossary OR in-line glossary elements if desired. Otherwise, defaults to a fully combined glossary.

pkgdown example: https://github.com/r-lib/pkgdown/blob/main/pkgdown/_pkgdown.yml#L50-L64

@batpigandme also had an internal suggestion for MDN macros and their glossaries/indices: https://developer.mozilla.org/en-US/docs/MDN/Structures/Macros/Commonly-used_macros

The RStudio Connect team and @aronatkins create a pseudo-glossary for the Connect configuration settings: https://docs.rstudio.com/connect/admin/appendix/configuration/#Server.Settings but the formatting has been a bit of frustration.

**Comment by SHogenboom** [comment]
This is exactly the feature request I wanted to post! 

I believe the latter option is the preferred solution - at least for books and other large projects. A specified glossary (e.g., `_glossary.yml`) with `terms` & `definitions` which upon rendering converts all terms to internal hyperlinks and adds a final `glossary` pages akin to what happens with the `references`.

I believe this method is preferred over manually adding glossary items because it prevents author(s) from forgetting to mark relevant `terms`. This becomes especially relevant when multiple authors work on different parts of a book/website or when chapters are added over time.

With regards to styling I could see two ideas for presenting the `definitions`: either with a hover effect (similar to rerefenced psyTeachR glossary package) and/or as a note in the side column.

Really hope this can become an internal feature of Quarto. For now, I'll use the very handy glossary package I did not know yet existed :)

**Comment by debruine** [comment]
@SHogenboom those are great ideas, but I have some caveats about auto-linking (based on a lot of experience implementing glossary links in educational materials): 

- The auto-linking of terms should let you add variants so you can link the same definition to different verb forms or plural and singular versions.
- It would be good to have a manual option in addition to auto. When I write educational materials, I might want to link the term "argument" the first few times I use it, or the first time each chapter, but probably not 10 times in 1 paragraph. 
- Auto-linking probably only makes sense in main text, not in tables or figure captions or code chunks. Working out the scope of where it is and isn't appropriate to auto-link might be more work than a manual demarcation option, like only link words marked by {{:term:}}.

P.S. I'm glad you think the glossary package will be helpful! I'd planned on sending it to CRAN after a little more development to make it easier to set up your own glossary website (it's currently pretty specialised for our PsyTeachR purposes), but I'm putting that plan on hold now until I find out what quarto is doing with built-in glossaries. I'd definitely add a way to easily make _glossary.yml files to it, if that's the direction quarto goes.

**Comment by SHogenboom** [comment]
@debruine you make some very good arguments against auto-linking! I understand now that manual option would be preferred. And if I really wanted to highlight a term across the entire project a search & replace would suffice (which excludes figures, tables, captions etc).

In the coming weeks I will likely adapt the `PsyTeachR` code so that it functions within the scope of a project with a csv or .RData file. And then also adjust the generated code to differ the output based on the document type. After all the hover over effect won't work in a PDF file and I also like the idea of showing definitions in the margins.

**Comment by krionical** [comment]
Is there anything further on the status of this?

**Comment by cscheid** [comment]
@krionical when there are changes you'll see them here. We appreciate the enthusiasm, but do note that because we try hard to keep up with the issues here, every such question like this is time we don't get to work on quarto :)

**Comment by krionical** [comment]
Sure thing. I'll check back in the year 2025. I'll go ahead and pencil it in right now.

**Comment by salim-b** [comment]
To potentially avoid over-engineering this glossary thing: Pandoc already supports [**definition lists**](https://pandoc.org/MANUAL.html#definition-lists), which for HTML output render to [description list (`<dl>`) elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl). I think it would really make sense *semantically* for Quarto to build upon this.

Currently, one has to manually add fragment identifiers to the definition terms to be referenceable in HTML output.  
Example `glossary.qmd`:

```md
# Glossary {.unnumbered}

[Term 1]{#term-id-1}

:   Definition 1... bla bli bla
    blup.

[Term 2]{#term-id-2}

:   Definition 2... bla bli bla
    blup.
```

With the above, we can link to glossary term 1 anywhere in our Quarto project via `bla bla [Term 1](glossary.qmd#term-id-1) bla bla`.

Now what's missing the most for a definition-list-based glossary is proper [cross reference](https://quarto.org/docs/authoring/cross-references.html) support with autonumbering that produces human-parseable navigation hints (especially relevant for PDF output). Of course, dedicated Quarto shortlink(s) to create such definition-list-cross-references would be nice.

**Comment by nichtich** [comment]
See <https://github.com/quarto-dev/quarto-cli/discussions/805#discussioncomment-3082107> for an implementation based on a filter.

**Comment by andrewpbray** [comment]
See #6646 for a barebones implementation based on a filter.

**Comment by debruine** [comment]
FYI, I've started working on a glossary lua extension that uses the same glossary file format and arguments as my [glossary R package](https://debruine.github.io/glossary/). It's pretty rough (I started learning lua yesterday), but does the basics. Feedback is very welcome.

<https://github.com/debruine/quarto-glossary>

**Comment by cscheid** [comment]
I'm sure this is slightly disappointing news for folks here, but we're pushing this to the v1.5 release. We very much will implement this, and the infrastructure is there. But we want to make sure we get a release out sooner rather than later, so I'm reassigning the milestone. Thanks for the continued patience!

**Comment by roaldarbol** [comment]
Since the v1.5 release has been released, is there a new timeline for it? I can tell it says "Future", which doesn't help say much (usually means "not something we're currently planning to work on in the foreseeable future"). This is not to pile on pressure, priorities are necessary and completely understandable, just seems many folks had been looking forward to this feature so maybe an update would be appreciated. 😊

**Comment by jameshowison** [comment]
Yes, would love an update here.  I use this for class pages while teaching and rebuilding the whole site on each edit is a pain.

Does anyone know what sort of features mean that `quarto preview` is insufficient and requires a full rebuild?  Is it just references?  Perhaps what I mean is: "How far does `quarto preview` get you towards incremental builds?

**Comment by roaldarbol** [comment]
@jameshowison Is your question specifically related to glossaries? It seems perhaps better tracked by the https://github.com/quarto-dev/quarto-cli/issues/6518 epic issue and the issues it tracks?

**Comment by mcanouil** [comment]
The team updates all issues/discussions when there is something to update about.
Asking for updates is asking the team to spend time answering "no, there is no update".

Thanks for your interest in this issue and others, but adding reactions instead of comment is more valuable as it helps the team change priorities eventually.
![Image](https://github.com/user-attachments/assets/d902d9ce-f7bd-42c6-8b0e-f67538bee6c5)

Thank you for your understanding.

**Comment by jameshowison** [comment]
> [@jameshowison](https://github.com/jameshowison) Is your question specifically related to glossaries? It seems perhaps better tracked by the [#6518](https://github.com/quarto-dev/quarto-cli/issues/6518) epic issue and the issues it tracks?

Eeek, you are right, my apologies, wrong issue!

**Comment by roaldarbol** [comment]
> The team updates all issues/discussions when there is something to update about. Asking for updates is asking the team to spend time answering "no, there is no update".
> 
> Thanks for your interest in this issue and others, but adding reactions instead of comment is more valuable as it helps the team change priorities eventually.

Fair enough, I'll keep that in mind; it's the only existing issue I've commented on, it's not like I'm spamming the issues for updates (I've opened new issues for bugs and feature requests, but I imagine that's still encouraged). 

I do think going from "We're going to support this in the next release" to "we don't know when we will support it" should merit a quick comment - it is after all the 2nd most requested feature (based on 👍 which you say prioritise efforts), and the highest which is not currently planned. 🤷

**Comment by cscheid** [comment]
> I do think going from "We're going to support this in the next release" to "we don't know when we will support it" 

That's not what happened, and we haven't implied that. Nothing in our plans has changed. We changed our workflow, that's it. In the past, we'd push all issues to the following release. That caused a consistent churn of "push to the next release" which takes time. We are just clarifying now which of those are ones we have plans for.

**Comment by roaldarbol** [comment]
That’s really really useful to know, thanks @cscheid! It’s exactly that kind of information that would be great to communicate; there’s no way we could no of that internal change otherwise.

**Comment by adulau** [comment]
Just a small note, we are planning to move [misp-book](https://github.com/MISP/misp-book/) away from honkit (a fork from the original open source version of git-book) but the glossary is the only missing feature. If a specific support is required for the Quarto team to include this feature, don't hesitate to let us know.

**Comment by krionical** [comment]
> [@krionical](https://github.com/krionical) when there are changes you'll see them here. We appreciate the enthusiasm, but do note that because we try hard to keep up with the issues here, every such question like this is time we don't get to work on quarto :)

Hey look, it's 2025! It's literally been almost 2 years! I'm glad I didn't ask any questions during this time so that features like this could be worked on! I went ahead and published that book and everything though, banking on the fact that probably the "your question is distracting me from important things I'm working on" was really "this isn't a priority for us". 

Just for a kind of quick comparison though, this guy wrote two novels in 18 months! https://lifehacker.com/how-i-wrote-two-full-length-novels-in-18-months-1794785678

**Comment by cscheid** [comment]
@krionical This comment violates our code of conduct. Please be more constructive in the future. If you repeat this behavior we will restrain you from further posting.

**Comment by krionical** [comment]
> @krionical This comment violates our code of conduct. Please be more constructive in the future. If you repeat this behavior we will restrain you from further posting.

I waited two years to comment as I said I would. You have not been constructive in this case, not me. From the point of posting, it is actually three years that you have done nothing where this feature is concerned. (or even offered an update) You are attacking your user base instead of doing the work. Metallica did something similar when it attacked its fans for downloading its music.

It might interest you to know that I've been doing some reporting concerning codes of conduct, and how they enable people to police "problem people" instead of actually doing the required work (which is coding) on a project. Whether you restrain me due to some interpretation of your code of conduct is irrelevant to me. It will doom Quarto to become irrelevant, however, which is too bad. (I hear grants at the government level are becoming a little harder to come by--especially in academic settings) Of course, if you then become jobless, it might help you focus on doing a job instead of accepting the fact that you and your project are failing to meet an obvious, required, goal for academic publishing--and you are willing to mute your users due to "perceived violations". Maybe Typst will take over where you obviously cannot go, or maybe the tried and true Latex, which never had a code of conduct for all those years, will assume the position where results on merit matter more than whatever political bullshit a project espouses is involved. 

So, go ahead, silence away, but don't whine about the consequences. You had an extended amount of time to do the right thing. 

It is ironic to write software to write books, though, and silence people because you don't like what they have to say because it points out the hypocrisy of your own position.

**Comment by cscheid** [comment]
I warned you. You're no longer allowed to post here.

**Comment by cscheid** [comment]
Everyone else: thank you for your patience and kindness, and we apologize for the noise. I _promise_ you this is on our radar. There's only so many of us and so many hours in the day...
